mmap-alloc/src/tests.rs

```diff
@@ -0,0 +1,584 @@
+
+        // NOTE: This test leaks memory, but it's only a page, so it doesn't really matter.
+        let ptr = map(pagesize(), PROT_READ_WRITE, false, None).unwrap();
+        unmap(ptr, 0);
+    }
+}
+
+#[test]
+#[should_panic]
+fn test_unmap_panic_unaligned() {
+    unsafe {
+        // Check that a non-page-aligned address causes unmap to panic.
+        unmap((pagesize() / 2) as *mut u8, pagesize());
+    }
+}
+
+#[cfg(not(windows))]
+#[cfg(not(feature = "test-no-std"))]
+#[bench]
+#[ignore]
+fn bench_large_map(b: &mut Bencher) {
+    // Determine the speed of mapping a large region of memory so that we can tune the timeout
+    // in test_map_non_windows.
+    b.iter(|| unsafe {
+        let ptr = map(1 << 29, PROT_READ_WRITE, false, None).unwrap();
+        unmap(ptr, 1 << 29);
+    })
+}
+
+// assert_block_perm asserts that every page in the block starting at ptr with the given size has
+// the given set of permissions.
+#[cfg(target_os = "linux")]
+fn assert_block_perm(ptr: *mut u8, size: usize, perm: Perm) {
+    let ptr = ptr as usize;
+    for block in perms() {
+        // If this Block overlaps with the block being queried, then since a Block can only have
+        // a single set of permissions, then their permissions must match.
+        let disjoint = (ptr + size) <= block.begin as usize || block.end as usize <= ptr;
+        if !disjoint {
+            assert_eq!(perm, block.perm);
+        }
+    }
+}
+
+// A Block represents a mapped block of memory with a particular set of permissions. The range is
+// inclusive/exclusive (i.e., [begin, end)).
+#[cfg(target_os = "linux")]
+#[derive(Debug)]
+struct Block {
+    begin: *mut u8,
+    end: *mut u8,
+    perm: Perm,
+}
+
+#[cfg(target_os = "linux")]
+#[test]
+fn test_perms_fn() {
+    perms();
+}
+
+// perms gets a list of the mapped blocks for this process and their associated permissions.
+#[cfg(target_os = "linux")]
+fn perms() -> Vec<Block> {
+    // perms works by parsing /proc/<pid>/maps, which has lines like
+    // 7f818a324000-7f818a326000 rw-p 001c4000 08:01 24360                      /lib/x86_64-linux-gnu/libc-2.23.so
+    // We care about the first three fields: the starting address, the ending address, and the
+    // permissions set.
+
+    use std::process::Command;
+    use std::collections::HashMap;
+
+    let mut map = HashMap::new();
+    map.insert(String::from("---p"), PROT_NONE);
+    map.insert(String::from("r--p"), PROT_READ);
+    map.insert(String::from("-w-p"), PROT_WRITE);
+    map.insert(String::from("--xp"), PROT_EXEC);
+    map.insert(String::from("rw-p"), PROT_READ_WRITE);
+    map.insert(String::from("r-xp"), PROT_READ_EXEC);
+    map.insert(String::from("-wxp"), PROT_WRITE_EXEC);
+    map.insert(String::from("rwxp"), PROT_READ_WRITE_EXEC);
+
+    let mut blocks = Vec::new();
+
+    let output = Command::new("cat")
+        .arg(format!("/proc/{}/maps", unsafe { libc::getpid() }))
+        .output()
+        .unwrap();
```

Comment:
Is there a reason to use cat vs opening the file ourselves with std::fs::File?

Reply:
This was adapted from an earlier version of the function whose job was to get the permissions for a _particular_ address block, and thus was a more complicated command involving a full sequence of commands chained together with pipes. I honestly didn't even notice what this ended up getting whittled down to - I'll change it.